src/Compilers/Core/Portable/Serialization/StreamObjectWriter.cs

```diff
@@ -1050,27 +1070,42 @@ private static Exception NoSerializationWriterException(string typeName)
         }
 
         // we have s_typeMap and s_reversedTypeMap since there is no bidirectional map in compiler
-        internal static readonly ImmutableDictionary<Type, EncodingKind> s_typeMap = ImmutableDictionary.CreateRange<Type, EncodingKind>(
-            new KeyValuePair<Type, EncodingKind>[]
-            {
-                KeyValuePair.Create(typeof(bool), EncodingKind.BooleanType),
-                KeyValuePair.Create(typeof(char), EncodingKind.Char),
-                KeyValuePair.Create(typeof(string), EncodingKind.StringType),
-                KeyValuePair.Create(typeof(sbyte), EncodingKind.Int8),
-                KeyValuePair.Create(typeof(short), EncodingKind.Int16),
-                KeyValuePair.Create(typeof(int), EncodingKind.Int32),
-                KeyValuePair.Create(typeof(long), EncodingKind.Int64),
-                KeyValuePair.Create(typeof(byte), EncodingKind.UInt8),
-                KeyValuePair.Create(typeof(ushort), EncodingKind.UInt16),
-                KeyValuePair.Create(typeof(uint), EncodingKind.UInt32),
-                KeyValuePair.Create(typeof(ulong), EncodingKind.UInt64),
-                KeyValuePair.Create(typeof(float), EncodingKind.Float4),
-                KeyValuePair.Create(typeof(double), EncodingKind.Float8),
-                KeyValuePair.Create(typeof(decimal), EncodingKind.Decimal),
-            });
-
-        internal static readonly ImmutableDictionary<EncodingKind, Type> s_reverseTypeMap 
-            = s_typeMap.ToImmutableDictionary(kv => kv.Value, kv => kv.Key);
+        // Note: these types are immutable.  However, for maxiumum perf we use mutable types because
+        // they are used in hotspots.
+        internal static readonly Dictionary<Type, EncodingKind> s_typeMap;
+
+        /// <summary>
+        /// Indexed by EncodingKind.
+        /// </summary>
+        internal static readonly Type[] s_reverseTypeMap;
```

Comment:
Wait a sec. Why is ImmutableArray slower than mutable here? That shouldn't be the case